## Determine major work destinations within census tracts in New Jersey

##### Which tracts have the highest number of jobs that people commute to from another county within the state using LODES database
##### To understand where residents within the state are employed

* Data source (2018): https://lehd.ces.census.gov/data/
* Variables:
* Residence Census Block Code (h_geocode)
* Workplace Census Block Code (w_geocode) 
* Total number of primary jobs (S000)
* Identify top 5 work destinations where NJ residents commute to work



In [1]:

import numpy as np
import pandas as pd
import matplotlib as mplib
import matplotlib.pyplot as plt

# For inline graphics
%matplotlib inline

In [2]:
pd.set_option('display.max_columns', None)

In [3]:
def get_nj_wac_main(year, state = "nj"):
    # state: string, two-letter code of state for which we want the data
    # year: int, the year we want to bring in data for
    
    base_url = 'https://lehd.ces.census.gov/data/lodes/LODES7/'
    file_specs = '{st}/wac/{st}_wac_SI03_JT05_{yr}.csv.gz'.format(st = state, yr = year)
    file_name = base_url + file_specs
    
    print("The URL for the file is at: " + file_name)
    output = pd.read_csv(file_name,compression='gzip')
    return(output)

In [4]:
df_wac_2018 = get_nj_wac_main(year =2018, state = 'nj')
df_wac_2018.head()

The URL for the file is at: https://lehd.ces.census.gov/data/lodes/LODES7/nj/wac/nj_wac_SI03_JT05_2018.csv.gz


w_geocode  C000  CA01  CA02  CA03  CE01  CE02  CE03  CNS01  CNS02  \
0  340010012002020   169     2   101    66     1     4   164      0      0   
1  340010024002005  1476    77   847   552     3    11  1462      0      0   
2  340010024002010     1     1     0     0     0     1     0      0      0   
3  340010024003012     8     0     4     4     0     0     8      0      0   
4  340010024003022     1     0     1     0     0     0     1      0      0   

   CNS03  CNS04  CNS05  CNS06  CNS07  CNS08  CNS09  CNS10  CNS11  CNS12  \
0      0      0      0      0      0      0      0      0      0      0   
1      0      0      0      0      0      0      0      0      0      0   
2      0      0      0      0      0      0      0      0      0      0   
3      0      0      0      0      0      0      8      0      0      0   
4      0      0      0      0      0      0      0      0      0      1   

   CNS13  CNS14  CNS15  CNS16  CNS17  CNS18  CNS19  CNS20  CR01  CR02  CR03  \
0      0      0      0      0      0      0      0    169   145    10     0   
1      0      0      0      0      0      0      0   1476  1157   121    13   
2      0      0      0      0      0      0      0      1     1     0     0   
3      0      0      0      0      0      0      0      0     7     1     0   
4      0      0      0      0      0      0      0      0     1     0     0   

   CR04  CR05  CR07  CT01  CT02  CD01  CD02  CD03  CD04  CS01  CS02  CFA01  \
0     6     0     8   162     7     0    30    51    86    95    74      0   
1   137     6    42  1405    71    15    94   308   982  1020   456      0   
2     0     0     0     1     0     0     0     0     0     1     0      0   
3     0     0     0     8     0     0     2     1     5     0     8      0   
4     0     0     0     1     0     0     0     0     1     1     0      0   

   CFA02  CFA03  CFA04  CFA05  CFS01  CFS02  CFS03  CFS04  CFS05  createdate  
0      0      0      0      0      0      0      0      0      0    20201120  
1      0      0      0      0      0      0      0      0      0    20201120  
2      0      0      0      0      0      0      0      0      0    20201120  
3      0      0      0      0      0      0      0      0      0    20201120  
4      0      0      0      0      0      0      0      0      0    20201120

In [5]:
def get_nj_rac_main(year, state = "nj"):
    # state: string, two-letter code of state for which we want the data
    # year: int, the year we want to bring in data for
    
    base_url = 'https://lehd.ces.census.gov/data/lodes/LODES7/'
    file_specs = '{st}/rac/{st}_rac_SI03_JT05_{yr}.csv.gz'.format(st = state, yr = year)
    file_name = base_url + file_specs
    
    print("The URL for the file is at: " + file_name)
    output = pd.read_csv(file_name,compression='gzip')
    return(output)

In [6]:
df_rac_2018 = get_nj_rac_main(year =2018, state = 'nj')
df_rac_2018.head()

The URL for the file is at: https://lehd.ces.census.gov/data/lodes/LODES7/nj/rac/nj_rac_SI03_JT05_2018.csv.gz


h_geocode  C000  CA01  CA02  CA03  CE01  CE02  CE03  CNS01  CNS02  \
0  340010001001009     1     1     0     0     0     0     1      0      0   
1  340010001002018     1     1     0     0     0     1     0      0      0   
2  340010002001010     1     0     1     0     0     0     1      0      0   
3  340010002002003     1     0     1     0     0     0     1      0      0   
4  340010002002012     1     0     0     1     0     0     1      0      0   

   CNS03  CNS04  CNS05  CNS06  CNS07  CNS08  CNS09  CNS10  CNS11  CNS12  \
0      0      0      0      0      0      0      0      0      0      0   
1      0      0      0      0      0      0      0      0      0      0   
2      0      0      0      0      0      0      0      0      0      0   
3      0      0      0      0      0      0      0      0      0      0   
4      0      0      0      0      0      0      0      0      0      0   

   CNS13  CNS14  CNS15  CNS16  CNS17  CNS18  CNS19  CNS20  CR01  CR02  CR03  \
0      0      0      0      0      0      0      0      1     0     0     0   
1      0      0      0      0      0      0      0      1     0     1     0   
2      0      0      0      0      0      0      0      1     0     0     0   
3      0      0      0      0      0      0      0      1     1     0     0   
4      0      0      0      0      0      0      0      1     1     0     0   

   CR04  CR05  CR07  CT01  CT02  CD01  CD02  CD03  CD04  CS01  CS02  \
0     1     0     0     1     0     0     0     0     0     0     1   
1     0     0     0     1     0     0     0     0     0     0     1   
2     1     0     0     1     0     0     0     0     1     1     0   
3     0     0     0     1     0     0     0     1     0     0     1   
4     0     0     0     1     0     0     0     0     1     0     1   

   createdate  
0    20201120  
1    20201120  
2    20201120  
3    20201120  
4    20201120

In [7]:
def get_geo_xwalk_main(state = "nj"):
    # state: string, two-letter code of state for which we want the data
    # year: int, the year we want to bring in data for
    
    base_url = 'https://lehd.ces.census.gov/data/lodes/LODES7/'
    file_specs = '{st}/{st}_xwalk.csv.gz'.format(st = state)
    file_name = base_url + file_specs
    
    print("The URL for the file is at: " + file_name)
    output = pd.read_csv(file_name,compression='gzip')
    return(output)

In [8]:
df_xwalk_2018 = get_geo_xwalk_main(state = 'nj')
df_xwalk_2018.head()

The URL for the file is at: https://lehd.ces.census.gov/data/lodes/LODES7/nj/nj_xwalk.csv.gz


/Users/NidhiTandon/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3361: DtypeWarning: Columns (27,28,29,37) have mixed types.Specify dtype option on import or set low_memory=False.
  if (await self.run_code(code, result,  async_=asy)):


tabblk2010  st stusps      stname    cty              ctyname  \
0  340090204001109  34     NJ  New Jersey  34009  Cape May County, NJ   
1  340090221011038  34     NJ  New Jersey  34009  Cape May County, NJ   
2  340210028004009  34     NJ  New Jersey  34021    Mercer County, NJ   
3  340210028004015  34     NJ  New Jersey  34021    Mercer County, NJ   
4  340210028004004  34     NJ  New Jersey  34021    Mercer County, NJ   

          trct               trctname          bgrp  \
0  34009020400     204 (Cape May, NJ)  340090204001   
1  34009022101  221.01 (Cape May, NJ)  340090221011   
2  34021002800        28 (Mercer, NJ)  340210028004   
3  34021002800        28 (Mercer, NJ)  340210028004   
4  34021002800        28 (Mercer, NJ)  340210028004   

                         bgrpname   cbsa               cbsaname  zcta  \
0     1 (Tract 204, Cape May, NJ)  36140         Ocean City, NJ  8270   
1  1 (Tract 221.01, Cape May, NJ)  36140         Ocean City, NJ  8251   
2        4 (Tract 28, Mercer, NJ)  45940  Trenton-Princeton, NJ  8629   
3        4 (Tract 28, Mercer, NJ)  45940  Trenton-Princeton, NJ  8629   
4        4 (Tract 28, Mercer, NJ)  45940  Trenton-Princeton, NJ  8629   

   zctaname    stplc stplcname     ctycsub                     ctycsubname  \
0    8270.0  9999999       NaN  3400974810   Upper township (Cape May, NJ)   
1    8251.0  9999999       NaN  3400945810  Middle township (Cape May, NJ)   
2    8629.0  9999999       NaN  3402129310  Hamilton township (Mercer, NJ)   
3    8629.0  9999999       NaN  3402129310  Hamilton township (Mercer, NJ)   
4    8629.0  9999999       NaN  3402129310  Hamilton township (Mercer, NJ)   

   stcd116 stcd116name  stsldl                        stsldlname  stsldu  \
0     3402       NJ-02   34001   General Assembly District 1, NJ   34001   
1     3402       NJ-02   34001   General Assembly District 1, NJ   34001   
2     3404       NJ-04   34014  General Assembly District 14, NJ   34014   
3     3404       NJ-04   34014  General Assembly District 14, NJ   34014   
4     3404       NJ-04   34014  General Assembly District 14, NJ   34014   

                     stslduname  stschool  \
0   State Senate District 1, NJ   3416650   
1   State Senate District 1, NJ   3410020   
2  State Senate District 14, NJ   3406540   
3  State Senate District 14, NJ   3406540   
4  State Senate District 14, NJ   3406540   

                            stschoolname  stsecon stseconname   trib tribname  \
0     Upper Township School District, NJ  9999999         NaN  99999      NaN   
1    Middle Township School District, NJ  9999999         NaN  99999      NaN   
2  Hamilton Township School District, NJ  9999999         NaN  99999      NaN   
3  Hamilton Township School District, NJ  9999999         NaN  99999      NaN   
4  Hamilton Township School District, NJ  9999999         NaN  99999      NaN   

      tsub  tsubname   stanrc  stanrcname  necta  nectaname  mil milname  \
0  9999999       NaN  9999999         NaN  99999        NaN  NaN     NaN   
1  9999999       NaN  9999999         NaN  99999        NaN  NaN     NaN   
2  9999999       NaN  9999999         NaN  99999        NaN  NaN     NaN   
3  9999999       NaN  9999999         NaN  99999        NaN  NaN     NaN   
4  9999999       NaN  9999999         NaN  99999        NaN  NaN     NaN   

      stwib                 stwibname   blklatdd   blklondd  createdate  
0  34340001  01 Atlantic-Cape May WIB  39.274472 -74.829342    20201120  
1  34340001  01 Atlantic-Cape May WIB  39.044120 -74.920564    20201120  
2  34340010             10 Mercer WIB  40.224083 -74.724779    20201120  
3  34340010             10 Mercer WIB  40.222765 -74.725446    20201120  
4  34340010             10 Mercer WIB  40.225823 -74.727454    20201120

In [9]:
def get_od_main(year, state = "nj"):
    # state: string, two-letter code of state for which we want the data
    # year: int, the year we want to bring in data for
    
    base_url = 'https://lehd.ces.census.gov/data/lodes/LODES7/'
    file_specs = '{st}/od/{st}_od_main_JT05_{yr}.csv.gz'.format(st = state, yr = year)
    file_name = base_url + file_specs
    
    print("The URL for the file is at: " + file_name)
    output = pd.read_csv(file_name,compression='gzip')
    return(output)

In [12]:
df_2018 = get_od_main(year =2018, state = 'nj')
df_2018.head()

The URL for the file is at: https://lehd.ces.census.gov/data/lodes/LODES7/nj/od/nj_od_main_JT05_2018.csv.gz


w_geocode        h_geocode  S000  SA01  SA02  SA03  SE01  SE02  SE03  \
0  340010012002020  340010001002018     1     1     0     0     0     1     0   
1  340010012002020  340010005001002     1     0     1     0     0     0     1   
2  340010012002020  340010005001003     1     0     1     0     0     0     1   
3  340010012002020  340010101022013     1     0     1     0     0     0     1   
4  340010012002020  340010102002023     1     0     0     1     0     0     1   

   SI01  SI02  SI03  createdate  
0     0     0     1    20201120  
1     0     0     1    20201120  
2     0     0     1    20201120  
3     0     0     1    20201120  
4     0     0     1    20201120

In [14]:
# Inner join on crosswalk data and O-D data 
df_final = pd.DataFrame({
    "tabblk2010": df_xwalk_2018.tabblk2010,
    "w_geocode": df_2018.w_geocode,
    "h_geocode": df_2018.h_geocode,
    "S000": df_2018.S000,
    "C000": df_rac_2018.C000,
    "trct": df_xwalk_2018.trct,
    "cty": df_xwalk_2018.cty
    
})
df_final 

tabblk2010     w_geocode     h_geocode  S000  C000         trct  \
0       340090204001109  3.400100e+14  3.400100e+14   1.0   1.0  34009020400   
1       340090221011038  3.400100e+14  3.400100e+14   1.0   1.0  34009022101   
2       340210028004009  3.400100e+14  3.400100e+14   1.0   1.0  34021002800   
3       340210028004015  3.400100e+14  3.400101e+14   1.0   1.0  34021002800   
4       340210028004004  3.400100e+14  3.400101e+14   1.0   1.0  34021002800   
...                 ...           ...           ...   ...   ...          ...   
169583  340299801001017           NaN           NaN   NaN   NaN  34029980100   
169584  340339900000009           NaN           NaN   NaN   NaN  34033990000   
169585  340339900000007           NaN           NaN   NaN   NaN  34033990000   
169586  340339900000004           NaN           NaN   NaN   NaN  34033990000   
169587  340339900000006           NaN           NaN   NaN   NaN  34033990000   

          cty  
0       34009  
1       34009  
2       34021  
3       34021  
4       34021  
...       ...  
169583  34029  
169584  34033  
169585  34033  
169586  34033  
169587  34033  

[169588 rows x 7 columns]

In [19]:
#pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [15]:
# create seperate home and work county fips columns as strings to compare with one another
df_final['cty_h'] = df_final['h_geocode'].apply(str).str.slice(0,4)
df_final['cty_w'] = df_final['w_geocode'].apply(str).str.slice(0,4)
df_final

tabblk2010     w_geocode     h_geocode  S000  C000         trct  \
0       340090204001109  3.400100e+14  3.400100e+14   1.0   1.0  34009020400   
1       340090221011038  3.400100e+14  3.400100e+14   1.0   1.0  34009022101   
2       340210028004009  3.400100e+14  3.400100e+14   1.0   1.0  34021002800   
3       340210028004015  3.400100e+14  3.400101e+14   1.0   1.0  34021002800   
4       340210028004004  3.400100e+14  3.400101e+14   1.0   1.0  34021002800   
...                 ...           ...           ...   ...   ...          ...   
169583  340299801001017           NaN           NaN   NaN   NaN  34029980100   
169584  340339900000009           NaN           NaN   NaN   NaN  34033990000   
169585  340339900000007           NaN           NaN   NaN   NaN  34033990000   
169586  340339900000004           NaN           NaN   NaN   NaN  34033990000   
169587  340339900000006           NaN           NaN   NaN   NaN  34033990000   

          cty cty_h cty_w  
0       34009  3400  3400  
1       34009  3400  3400  
2       34021  3400  3400  
3       34021  3400  3400  
4       34021  3400  3400  
...       ...   ...   ...  
169583  34029   nan   nan  
169584  34033   nan   nan  
169585  34033   nan   nan  
169586  34033   nan   nan  
169587  34033   nan   nan  

[169588 rows x 9 columns]

In [16]:
df_final.dropna(inplace=True)

In [17]:
df_final['w_geocode'] = df_final['w_geocode'].astype('int64')

In [18]:
df_final['h_geocode'] = df_final['h_geocode'].astype('int64')

In [19]:
df_final.head()

tabblk2010        w_geocode        h_geocode  S000  C000         trct  \
0  340090204001109  340010012002020  340010001002018   1.0   1.0  34009020400   
1  340090221011038  340010012002020  340010005001002   1.0   1.0  34009022101   
2  340210028004009  340010012002020  340010005001003   1.0   1.0  34021002800   
3  340210028004015  340010012002020  340010101022013   1.0   1.0  34021002800   
4  340210028004004  340010012002020  340010102002023   1.0   1.0  34021002800   

     cty cty_h cty_w  
0  34009  3400  3400  
1  34009  3400  3400  
2  34021  3400  3400  
3  34021  3400  3400  
4  34021  3400  3400

In [20]:

# filter blocks that have different work and home county
df_filtered = df_final[(df_final['cty_h'] != df_final['cty_w'])]


In [21]:
df_filtered.head()

tabblk2010        w_geocode        h_geocode  S000  C000  \
119  340057023001031  340010012002020  340110303003028   1.0   1.0   
120  340057024001041  340010012002020  340110304005045   1.0   1.0   
121  340090204001148  340010012002020  340110406002002   1.0   3.0   
122  340090207002022  340010012002020  340110407002002   1.0   1.0   
123  340090207002012  340010012002020  340155002053044   1.0   1.0   

            trct    cty cty_h cty_w  
119  34005702300  34005  3401  3400  
120  34005702400  34005  3401  3400  
121  34009020400  34009  3401  3400  
122  34009020700  34009  3401  3400  
123  34009020700  34009  3401  3400

In [22]:
# aggregate blocks to tracts and sum S000 column
data_agg = df_filtered.groupby(['trct', 'cty'], as_index=False)[['S000']].sum()
data_agg

trct    cty  S000
0    34005702300  34005   1.0
1    34005702400  34005   1.0
2    34009020400  34009   2.0
3    34009020700  34009   2.0
4    34017000100  34017   4.0
..           ...    ...   ...
207  34035053707  34035   2.0
208  34035053805  34035   1.0
209  34035053904  34035   1.0
210  34035054100  34035  29.0
211  34035054202  34035   2.0

[212 rows x 3 columns]

In [23]:
# sort by S000 column descending
data_agg = data_agg.sort_values('S000', ascending=False)

In [24]:
data_agg.head(25)

trct    cty   S000
164  34019010400  34019  141.0
167  34019010701  34019   98.0
166  34019010600  34019   95.0
165  34019010500  34019   94.0
178  34019011500  34019   92.0
162  34019010200  34019   85.0
87   34017012700  34017   74.0
179  34019011600  34019   60.0
197  34035052700  34035   60.0
173  34019011002  34019   56.0
194  34035052400  34035   55.0
103  34017014600  34017   54.0
175  34019011301  34019   54.0
181  34019011800  34019   51.0
83   34017011500  34017   49.0
163  34019010300  34019   47.0
70   34017010100  34017   45.0
172  34019011001  34019   41.0
97   34017014102  34017   40.0
52   34017006100  34017   40.0
176  34019011302  34019   37.0
45   34017005400  34017   37.0
76   34017010800  34017   36.0
155  34017019800  34017   35.0
156  34017019900  34017   34.0

In [25]:
data_agg['%'] = ((data_agg['S000'] / data_agg['S000'].sum())*100).round(2).astype(str) + '%'

In [26]:
data_agg.head()

trct    cty   S000      %
164  34019010400  34019  141.0  3.97%
167  34019010701  34019   98.0  2.76%
166  34019010600  34019   95.0  2.68%
165  34019010500  34019   94.0  2.65%
178  34019011500  34019   92.0  2.59%

In [27]:
# sort by S000 column descending
data_agg = data_agg.sort_values('%', ascending=False)

In [28]:
data_agg.head(25)

trct    cty   S000      %
164  34019010400  34019  141.0  3.97%
167  34019010701  34019   98.0  2.76%
166  34019010600  34019   95.0  2.68%
165  34019010500  34019   94.0  2.65%
178  34019011500  34019   92.0  2.59%
162  34019010200  34019   85.0  2.39%
87   34017012700  34017   74.0  2.08%
179  34019011600  34019   60.0  1.69%
197  34035052700  34035   60.0  1.69%
173  34019011002  34019   56.0  1.58%
194  34035052400  34035   55.0  1.55%
103  34017014600  34017   54.0  1.52%
175  34019011301  34019   54.0  1.52%
181  34019011800  34019   51.0  1.44%
83   34017011500  34017   49.0  1.38%
163  34019010300  34019   47.0  1.32%
70   34017010100  34017   45.0  1.27%
172  34019011001  34019   41.0  1.15%
97   34017014102  34017   40.0  1.13%
52   34017006100  34017   40.0  1.13%
176  34019011302  34019   37.0  1.04%
45   34017005400  34017   37.0  1.04%
76   34017010800  34017   36.0  1.01%
155  34017019800  34017   35.0  0.99%
156  34017019900  34017   34.0  0.96%

#### County Codes for NJ

* 34001	Atlantic	NJ
* 34003	Bergen	NJ
* 34005	Burlington	NJ
* 34007	Camden	NJ
* 34009	Cape May	NJ
* 34011	Cumberland	NJ
* 34013	Essex	NJ
* 34015	Gloucester	NJ
* 34017	Hudson	NJ
* 34019	Hunterdon	NJ
* 34021	Mercer	NJ
* 34023	Middlesex	NJ
* 34025	Monmouth	NJ
* 34027	Morris	NJ
* 34029	Ocean	NJ
* 34031	Passaic	NJ
* 34033	Salem	NJ
* 34035	Somerset	NJ
* 34037	Sussex	NJ
* 34039	Union	NJ


### The top 5 county destinations where NJ residents commute to work are Hunterdon, Hudson, Somerset, Morris, and Mercer